cloudigrade/api/views.py

```diff
@@ -75,49 +75,71 @@ def list(self, *args, **kwargs):
         return Response(response)
 
 
-@method_decorator(transaction.non_atomic_requests, name="dispatch")
-class DailyConcurrentUsageViewSet(viewsets.GenericViewSet, mixins.ListModelMixin):
-    """
-    Generate report of concurrent usage within a time frame.
-
-    This viewset has to be wrapped in a non_atomic_requests decorator. DRF by default
-    runs viewset methods inside of an atomic transaction. But in this case we have to
-    create ConcurrentUsageCalculationTask to schedule jobs even when we raise a
-    ResultsUnavailable 425 exception.
-    """
+class SharedDailyConcurrentUsageViewSet(viewsets.GenericViewSet, mixins.ListModelMixin):
+    """Generate report of concurrent usage within a time frame."""
 
     schema = schemas.ConcurrentSchema(tags=["api-v2"])
     serializer_class = serializers.DailyConcurrentUsageSerializer
 
-    def get_queryset(self):  # noqa: C901
+    def get_queryset(
+        self,
+        default_start_date=None,
+        default_end_date=None,
+        latest_start_date=None,
+        latest_end_date=None,
```

Comment:
I believe Django apps tend not to define and use custom arguments in functions like `get_queryset` because that could conflict with how the default view classes load them, which may not expose an interface to provide additional arguments. Although the implementation here probably works just fine in our case, I don't know if it looks like a Django "best practice".

Instead of overriding the call to `get_queryset` in the child classes, what would you think about the following ideas?

We could define separate helper methods (_maybe_ decorated as properties, but I don't have as strong feelings about that here) that `get_queryset` would use for each of the variables that differs between the child classes. Then the parent `get_queryset` here would call _those_ methods, and any child classes would _only_ implement those methods, not the `get_queryset` method itself. 

Also, we could drop the `SharedDailyConcurrentUsageViewSet` distinction entirely since it probably doesn't need to exist. I think if we were to diagram out the use and inheritance tree, `SharedDailyConcurrentUsageViewSet` looks and behaves more like an abstract base class since nothing instantiates it directly and only its child classes are instantiated. We could flatten the relationship a bit so that only `DailyConcurrentUsageViewSet` and `InternalDailyConcurrentUsageViewSet` exists, with the latter being a child of the former and the latter only implementing those extra getter/property methods.

Combining those ideas, the implementation could look like this pseudocode:

```
class DailyConcurrentUsageViewSet(viewsets.GenericViewSet, mixins.ListModelMixin):
  @property
  def latest_start_date(self):
    return get_today() - timedelta(days=1)

  def get_queryset(self):
    ...
    latest_start_date = self.latest_start_date
    ...

class InternalDailyConcurrentUsageViewSet(DailyConcurrentUsageViewSet):
  @property
  def latest_start_date(self):
    return get_today()
```

Reply:
yea, that was one of the possibilities (class methods instead of parameters), I'll give it a whirl. 👍